chapters/operatorsandexpressions.tex

```diff
@@ -537,7 +537,8 @@ \subsection{Derivative and Special Purpose Operators with Function Syntax}\doubl
 \lstinline!der!(expr) &
 The time derivative of \lstinline!expr!. If the expression \lstinline!expr! is a
 scalar it needs to be a subtype of Real. The expression and all its
-subexpressions must be differentiable. If \lstinline!expr! is an array, the operator
+time-varying subexpressions must be continuous and semi-differentiable.
+If \lstinline!expr! is an array, the operator
```

Comment:
When replacing _differentiable_ with something like _semi-differentiable_, I think we should think carefully about whether it is _semi-differentiable_ or _right differentiable_ that we want.  Getting this right might help interfacing Modelica signals with environments that don't share the idea of super-dense time, and a signal is required to have a single value at any point in time.

What would we loose by saying _right differentiable_ instead of _semi-differentiable_?

Reply:
Other than that, I'm in favor of a change in this direction, as it brings down #2423 to getting rid of the _continuous_.

Reply:
To me "subtype of Real" is both easy-to-understand (even if redundant) and currently needed, because we currently don't allow der of record expressions/variables.

Regarding right-differentiable vs. semi-differentiable I was unsure about whether it differed between the following two expressions: `noEvent(if time<0 then 0 else time)` and `noEvent(if time<=0 then 0 else time)`; and thus preferred the slightly more general semi-differentiable.

Reply:
Perhaps saying _scalar_ is more to the point then?

Reply:
You are right that the right vs semi-differentiable is a tricky question.  I think my preference for right differentiability comes from the way we store results, where one can't tell the difference between `if time < 0 then -1 else 1` and `if time <= 0 then -1 else 1`.  For some reason I've gotten used to always selecting the right value at an event when asked to read off a single value at the time of an event, and so the trajectories have ended up being right continuous in my mind.

I'd be interested to see a simulation result format where the kind of continuity is specified for each event time.  Given that, I think it would be clear that semi-differentiable is the concept to use.

Reply:
> Perhaps saying scalar is more to the point then?

Sorry, I don't see the line it is referring to, as der can be applied to scalars and arrays of reals.

Reply:
Sorry, I forgot that this was in a context already restricted to scalars, but then I don't understand this:
> currently needed, because we currently don't allow der of record expressions/variables

How about this:
> If the expression \lstinline!expr! is a scalar, the expression and all its time-varying subexpressions must be continuous and semi-differentiable.